### PR TITLE
Static argument descriptor

### PR DESCRIPTION
**Pull request by chrisseaton**
This is all the static information about arguments needed to be sent from the calleer to the callee in order to implement call-target-agnostic keyword arguments. It's also the most wide-ranging change needed to the calling convention.

Co-authored with @wildmaples.

### CODE REVIEW

File: spec/tags/truffle/arguments_descriptor_tags.txt [hunk 1]
```diff
@@ -0,0 +1,2 @@
+fails:Arguments descriptors work for a mixture of arguments
+fails:Arguments descriptors work for a call like our Struct.new
```
Comment (chrisseaton) [inline spec/tags/truffle/arguments_descriptor_tags.txt:2]: These two are legitimately failing specs for keyword argument behaviour in general - nothing to do with the descriptors.

File: spec/truffle/arguments_descriptor_spec.rb [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+  end
+end
+
+describe "Arguments descriptors" do
+  it "are empty for simple calls" do
+    ArgumentsDescriptorSpecs.no_kws(1, 2).should == [1, 2, [], ArgumentsDescriptorSpecs.tr? ? [1, 2]: []]
+  end
+
+  it "contain keyword arguments" do
+    ArgumentsDescriptorSpecs.only_kws(a: 1, b: 2).should == [1, 2, ArgumentsDescriptorSpecs.tr? ? [:a, :b, 0, true] : [], ArgumentsDescriptorSpecs.tr? ? [{a: 1, b: 2}] : []]
```
Comment (eregon) [inline spec/truffle/arguments_descriptor_spec.rb:93]: This is not readable, how about always ignoring the last 2 values on CRuby?
Could be like:
```ruby
engine = TruffleRuby ? (0..-1) : (0..-3)
ArgumentsDescriptorSpecs.only_kws(a: 1, b: 2)[engine].should == [1, 2, [:a, :b, 0, true], [{a: 1, b: 2}]][engine]
```

Alternatively, since this is a spec/truffle there is no need to make it work on CRuby, so maybe just don't try to run this on CRuby?

Comment (eregon) [reply spec/truffle/arguments_descriptor_spec.rb:93]: Could also be:
```ruby
check ArgumentsDescriptorSpecs.only_kws(a: 1, b: 2), [1, 2, [:a, :b, 0, true], [{a: 1, b: 2}]]
```
and define `check` in the `describe`

Comment (chrisseaton) [reply spec/truffle/arguments_descriptor_spec.rb:93]: > so maybe just don't try to run this on CRuby?

It also checks the actual handling of the arguments (which is why two are tagged on TruffleRuby.) But I could split them up.

Comment (eregon) [reply spec/truffle/arguments_descriptor_spec.rb:93]: It's probably useful to let it run on CRuby to check expectations are correct and match CRuby behavior, but it needs less noise.

File: src/main/java/org/truffleruby/language/arguments/ArgumentsDescriptor.java [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+ * redistribute it and/or modify it under the terms of the:
+ *
+ * Eclipse Public License version 2.0, or
+ * GNU General Public License version 2, or
+ * GNU Lesser General Public License version 2.1.
+ */
+package org.truffleruby.language.arguments;
+
+/** An arguments descriptor is an object that describes something about the arguments being passed. Argument descriptors
+ * are stable, immutable objects, suitable to be cached and guarded by identity. They'd normally convey some kind of
```
[the remark this answers is not in the document]
Comment (chrisseaton) [reply src/main/java/org/truffleruby/language/arguments/ArgumentsDescriptor.java:13]: 'They'd' is a contraction of 'they would'.

I've clarified that it's about the values of arguments passed.

File: spec/truffle/arguments_descriptor_spec.rb (hunk 2 above)
Comment (chrisseaton) [reply spec/truffle/arguments_descriptor_spec.rb]: Split it up to test one part per line.

**Comment by chrisseaton** [comment]
@eregon can you help me pair on some specs to exercise the questions you've got for the last three comments?

File: src/main/java/org/truffleruby/core/module/ModuleNodes.java [hunk 4]
```diff
@@ -800,7 +801,8 @@ protected Object classExec(RubyModule self, Object[] args, RubyProc block) {
                     new FixedDefaultDefinee(self),
                     block.declarationContext.getRefinements());
 
-            return callBlockNode.executeCallBlock(declarationContext, block, self, block.block, args);
+            return callBlockNode.executeCallBlock(declarationContext, block, self, block.block,
+                    EmptyArgumentsDescriptor.INSTANCE, args);
```
Comment (eregon) [inline src/main/java/org/truffleruby/core/module/ModuleNodes.java:805]: Since this method delegates the arguments to the block it should pass the existing descriptor, isn't it?

File: src/main/java/org/truffleruby/language/arguments/ArgumentsDescriptor.java [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+ * redistribute it and/or modify it under the terms of the:
+ *
+ * Eclipse Public License version 2.0, or
+ * GNU General Public License version 2, or
+ * GNU Lesser General Public License version 2.1.
+ */
+package org.truffleruby.language.arguments;
+
+/** An arguments descriptor is an object that describes something about the arguments being passed, regardless of what
+ * the caller expects or accepts. Argument descriptors are stable, immutable objects, suitable to be cached and guarded
```
Comment (eregon) [inline src/main/java/org/truffleruby/language/arguments/ArgumentsDescriptor.java:13]:
```suggestion
 * the callee expects or accepts. Argument descriptors are stable, immutable objects, suitable to be cached and guarded
```

File: src/main/java/org/truffleruby/language/arguments/ArgumentsDescriptorManager.java [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+            }
+            Key that = (Key) o;
+            return keywordHashIndex == that.keywordHashIndex && keywordHashElidable == that.keywordHashElidable &&
+                    Arrays.equals(keywordArgumentNames, that.keywordArgumentNames);
+        }
+
+        @Override
+        public int hashCode() {
+            int result = Objects.hash(keywordHashIndex, keywordHashElidable);
+            result = 31 * result + Arrays.hashCode(keywordArgumentNames);
```
Comment (eregon) [inline src/main/java/org/truffleruby/language/arguments/ArgumentsDescriptorManager.java:68]: Could you move the `Arrays.hashCode(keywordArgumentNames)` as the 3rd arg to Objects.hash? That way no need to hardcode 31 here

File: src/main/java/org/truffleruby/language/arguments/EmptyArgumentsDescriptor.java [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+ * code is released under a tri EPL/GPL/LGPL license. You can use it,
+ * redistribute it and/or modify it under the terms of the:
+ *
+ * Eclipse Public License version 2.0, or
+ * GNU General Public License version 2, or
+ * GNU Lesser General Public License version 2.1.
+ */
+package org.truffleruby.language.arguments;
+
+/** An arguments descriptor for where we're providing no information about the value of arguments being passed. */
```
Comment (eregon) [inline src/main/java/org/truffleruby/language/arguments/EmptyArgumentsDescriptor.java:12]: That sounds like `EmptyArgumentsDescriptor` can be used for all call sites, and receiving an empty EmptyArgumentsDescriptor doesn't imply there were no kwargs? I think that can't work, `EmptyArgumentsDescriptor` must mean no kwargs were passed.

**Comment by eregon** [comment]
I noticed the `super` specs don't pass, they only passed before because the specs didn't actually use `super`

**Comment by eregon** [comment]
> @eregon can you help me pair on some specs to exercise the questions you've got for the last three comments?

I'm not sure I can help much, but yes some specs would likely help to check the cases are correctly detected, e.g. mixes of `positional, *rest, literal_kw: value, **kw, &block`.
I think that code needs to be rewritten from scratch, the logic is too hard to follow and it's probably also incorrect.

File: src/main/java/org/truffleruby/language/arguments/EmptyArgumentsDescriptor.java (hunk 7 above)
Comment (chrisseaton) [reply src/main/java/org/truffleruby/language/arguments/EmptyArgumentsDescriptor.java:12]: Currently it means no information was provided. When we change in subsequent PRs to add other constraints or guarantees I'll update the comment. But at the moment it's just the abscence of optional information.

File: src/main/java/org/truffleruby/language/arguments/ArgumentsDescriptorManager.java (hunk 6 above)
Comment (eregon) [reply src/main/java/org/truffleruby/language/arguments/ArgumentsDescriptorManager.java:68]: I always found the IDE-generated hashCode/equals to be quite obtuse and hard to read, I think worth fixing.

File: src/main/java/org/truffleruby/language/arguments/EmptyArgumentsDescriptor.java (hunk 7 above)
Comment (eregon) [reply src/main/java/org/truffleruby/language/arguments/EmptyArgumentsDescriptor.java:12]: I feel this constraint is too weak. Either we give a proper descriptor in all places, or we're happy with just `EmptyArgumentsDescriptor` everywhere and add `NonEmptyArgumentsDescriptor` later.
A mix doesn't make sense to me, we'd try to figure the correct descriptor, but only in some arbitrary cases.

Comment (chrisseaton) [reply src/main/java/org/truffleruby/language/arguments/EmptyArgumentsDescriptor.java:12]: In this PR, descriptors are optional, but tested, so we're happy to do whatever.

In a future PR, descriptors will be required to understand the keyword arguments, if the caller wants to send them flattened.

If the sender doesn't or can't send them flattened, for whatever reason, it can send them with an empty descriptor even though there's keyword arguments.

(May change to required as we fix the last of the Ruby 3 semantics.)

So I think the comment is correct now, and a mix is ok, but that may be tightened in the future, in which case I'll change the comment to match.

Comment (eregon) [reply src/main/java/org/truffleruby/language/arguments/EmptyArgumentsDescriptor.java:12]: > If the sender doesn't or can't send them flattened, for whatever reason, it can send them with an empty descriptor even though there's keyword arguments.
> (May change to required as we fix the last of the Ruby 3 semantics.)

How can one possibly know (in the callee) if it was a positional hash or kwargs passed to a method?
E.g. `def m(*args, **kwargs)` called with `m({a: 1})` vs `m(a: 1)` or potentially delegated multiple times through `(*a, **kw)`.
I believe the descriptor must track this (it could be a separate bit somewhere, but what's the point?), and `EmptyArgumentsDescriptor` must therefore means "no kwargs".
